src/plan/sticky/immix/gc_work.rs

```diff
@@ -1,16 +1,20 @@
 use crate::policy::gc_work::TraceKind;
 use crate::policy::gc_work::TRACE_KIND_TRANSITIVE_PIN;
 use crate::scheduler::gc_work::PlanProcessEdges;
+use crate::scheduler::gc_work::UnsupportedProcessEdges;
 use crate::{plan::generational::gc_work::GenNurseryProcessEdges, vm::VMBinding};
 
 use super::global::StickyImmix;
 
-pub struct StickyImmixNurseryGCWorkContext<VM: VMBinding>(std::marker::PhantomData<VM>);
-impl<VM: VMBinding> crate::scheduler::GCWorkContext for StickyImmixNurseryGCWorkContext<VM> {
+pub struct StickyImmixNurseryGCWorkContext<VM: VMBinding, const KIND: TraceKind>(
+    std::marker::PhantomData<VM>
+);
+
+impl<VM: VMBinding, const KIND: TraceKind> crate::scheduler::GCWorkContext for StickyImmixNurseryGCWorkContext<VM, KIND> {
     type VM = VM;
     type PlanType = StickyImmix<VM>;
-    type ProcessEdgesWorkType = GenNurseryProcessEdges<VM, Self::PlanType>;
-    type TPProcessEdges = GenNurseryProcessEdges<VM, Self::PlanType>;
+    type ProcessEdgesWorkType = GenNurseryProcessEdges<VM, Self::PlanType, KIND>;
+    type TPProcessEdges = UnsupportedProcessEdges<VM>;
```

Comment:
That breaks Ruby, too.  Basically if you set `TPProcessEdges` to `UnsupportedProcessEdges`, any unmovable roots (delivered via `ProcessEdgesWorkRootsWorkFactory::create_process_pinning_roots_work`) will create an work packet that uses `UnsupportedProcessEdges` to trace the roots, which crashes.

Suggested changes:

```suggestion
    type ProcessEdgesWorkType = GenNurseryProcessEdges<VM, Self::PlanType, DEFAULT_TRACE>;
    type TPProcessEdges = GenNurseryProcessEdges<VM, Self::PlanType, TRACE_KIND_TRANSITIVE_PIN>;
```

And you can remove the generic parameter `<KIND>` from `StickyImmixNurseryGCWorkContext`, too.

I think the type member `type TPProcessEdges` is misleading.  It's not really related to "transitive pinning" per se.  The semantics is:

-    If an edge is updateable (i.e. its target is allowed to move), the GC should create a `ProcessEdgesWorkType` to process that edge.
-    If an edge cannot be updated (i.e. its target must not move), the GC should create a `TPProcessEdges` to process that edge.

What really happens is that when creating root packets, there are three options:
| method | process edges from root to object using | process edges between objects using |
|---|---|---|
| `create_process_edge_roots_work` | `ProcessEdgesWorkType` | `ProcessEdgesWorkType` |
| `create_process_pinning_roots_work` | `TPProcessEdges` | `ProcessEdgesWorkType` |
| `create_process_tpinning_roots_work` | `TPProcessEdges` | `TPProcessEdges` |

The kind of root selects the strategy of choosing which packets to use.  Non-transitive pinning roots process roots using `TPProcessEdges` but after that everything else can be processed with `ProcessEdgesWorkType`.  Transitive pinning is implemented with forcing all edges reachable from those roots to be traced using `TPProcessEdges`.

So the difference is just whether the edge can be updated.  I think there is much room for refactoring.

Reply:
Ohkay -- I should probably have tested that one-line change with ART before making this PR. It probably worked in my fork because I implemented `TPProcessEdges` instead of using the "unsupported" one.

This is _very_ confusing. I re-read the `create_process_pinning_roots_work` function and it's difficult to reason where the type parameters `I` and `E` are actually coming from and what they will be when using pinning roots vs transitively pinning roots. Your explanation makes sense but this is nowhere in the doc comment for the `ProcessEdgesWorkRootsWorkFactory` type.